### PR TITLE
Log offending non-positive value

### PR DESCRIPTION
**Pull request by frankmcsherry**
We leave early warning signs of defective data when we see non-positive accumulations in `reduce`, but we did not report the offending value. This PR modifies that complaint to also log the key, value, and count which led to the complaint.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/2991)

### CODE REVIEW

File: src/dataflow/render/reduce.rs [hunk 1]
```diff
@@ -483,11 +483,15 @@ where
...
         .reduce_named("ReduceHierarchical", {
-            move |_key, source, target| {
+            move |key, source, target| {
                 // Should negative accumulations reach us, we should loudly complain.
                 if source.iter().any(|(_val, cnt)| cnt <= &0) {
-                    // TODO(frank): Consider reporting the actual offending data.
-                    log::error!("Negative accumulation in ReduceHierarchical");
+                    for (val, cnt) in source.iter() {
+                        if cnt <= &0 {
+                            // XXX: This reports user data, which we perhaps should not do!
```
Comment (umanwizard) [inline src/dataflow/render/reduce.rs:491]: the wording here could be taken the wrong way IMO, like "YOLO, we haven't thought about privacy!"

Maybe something like "// TODO - make printing of user data configurable" would be better? Use your judgment I guess

Comment (frankmcsherry) [reply src/dataflow/render/reduce.rs:491]: I copy/pasted the other place in the file where we do this. I wanted something more scary sounding than "TODO" and "XXX" was what was recommended. Up for any language, but I want it to be as unmissable as possible.

Comment (quodlibetor) [reply src/dataflow/render/reduce.rs:491]: we could potentially have a string that we log in the output, to make it easier for customers to ensure that they know what they're sending us, e.g:

```rust
log::error!("[customer-data] Non-positive accumulation in ReduceHierarchical: key: {:?}\tvalue: {:?}\tcount: {:?}", key, val, cnt);
```

**Comment by frankmcsherry** [comment]
I like that suggestion! Will implement, and then merge.

**Pull request merged**